### ISSUE TITLE
Deterministic optimization routines (fminunc / fmincon)

### ISSUE THREAD

**Issue by rfeinman** [pytorch/pytorch#55279]
## 🚀 Feature
It would be great to have a library of optimization routines for the deterministic setting (a la `scipy.optimize`) using PyTorch autograd mechanics. I have written a [prototype library](https://github.com/rfeinman/pytorch-minimize) of PyTorch-based function minimizers and I'd like to guage the interest of the pytorch community. This could perhaps be a submodule of `optim`, or could be included elsewhere.

## Motivation

For a variety of optimization problems, deterministic (or "full-batch") routines such as Newton and Quasi-Newton methods are preferrable over vanilla gradient descent. Although PyTorch offers many routines for stochastic optimization, utilities for deterministic optimization are scarce; only L-BFGS is included in the `optim` package, and it's modified for mini-batch training.

MATLAB and SciPy are the industry standards for deterministic optimization. These libraries have a comprehensive set of routines; however, automatic differentiation is not supported. Therefore, the user must specify 1st- and 2nd-order gradients explicitly (they must be known) or use finite-difference approximations. This limits the applications considerably.

It would be highly valuable to have a new deterministic optimization library with analytical gradients via PyTorch's autograd.

## Pitch

See complete pitch at [pytorch-minimize](https://github.com/rfeinman/pytorch-minimize).

My current library offers 4 different routines for unconstrained minimization: BFGS, L-BFGS, Newton Conjugate Gradient, Newton Exact. A complete description of these algorithms is provided in the readme. Constrained minimization is not currently implemented, but could perhaps be included in the future.

```python
import torch
from fmin import minimize  # <- perhaps from torch.optim.fmin import minimize

def rosen(x):
    return torch.sum(100*(x[..., 1:] - x[..., :-1]**2)**2 
                     + (1 - x[..., :-1])**2)

# initial point
x0 = torch.tensor([1., 8.])

# BFGS
result = minimize(rosen, x0, method='bfgs')

# Newton Conjugate Gradient
result = minimize(rosen, x0, method='newton-cg')
```

**Comment by rfeinman** [comment]
Update: I've added a preliminary _constrained_ optimizer to the proposed library. I included a [tutorial](https://github.com/rfeinman/pytorch-minimize/blob/master/examples/constrained_optimization_adversarial_examples.ipynb) showing how to find an optimal adversarial perturbation subject to perturbation norm constraint.

**Comment by rfeinman** [comment]
@mruberry @mrshenli - I had an alternative idea that I want to run by you. 

I realize that adding a new "minimize" module could be difficult to coordinate since the schematic differs considerably from existing torch optimizers. As an alternative, I wrote a new ["Minimize"](https://github.com/rfeinman/pytorch-minimize/blob/master/fmin/optim.py) class that inherits from `optim.Optimizer` and supports parameter specifications just like existing optimizers. At the moment, the optimizer wraps `scipy.optimize.minimize` and supports all of its various [solver methods](https://docs.scipy.org/doc/scipy/reference/generated/scipy.optimize.minimize.html#scipy.optimize.minimize). Like `optim.LBFGS`, Minimize needs to re-evaluate the objective multiple times, so a closure callable must be provided.

Eventually it would be ideal to use custom torch implementations of the underlying solver algorithms and remove the scipy dependency. At the moment, Minimize interacts with scipy behind the scenes, so the front-end schematic is exactly as other optimizers (parameters are torch.Tensor or nn.Parameter, objectives use torch, etc.). CUDA is supported but will only be used for evaluating the objective and gradient; other numerical computations performed by the underlying solvers are on CPU (with numpy arrays).

**Comment by mruberry** [comment]
Hey @rfeinman! This is a cool idea and I'd like to discuss it with more of the PyTorch team.

One question we always have, however, with possible extensions is: could the desired functionality be provided as a library on top of PyTorch, or does it need to be included in PyTorch itself? Are there advantages to including this in PyTorch that may not be obvious?

**Comment by rfeinman** [comment]
@mruberry valid question. Deterministic optimization is a well-established area of machine learning, and I feel that PyTorch users could get a lot out of having these tools in the core library. I've seen a number of forum posts and tickets asking about the Conjugate Gradient optimizer, which is mentioned in the `torch.optim` documentation page but not actually implemented (I think it was previously available in torch, but never ported to pytorch?). Clearly there seems to be an interest in optimization routines of this family.

Although PyTorch is primarily a neural network library, people now use it for a range of general-purpose scientific computing applications, especially those which can be accelerated by GPU. In addition to supporting the neural network applications that I mentioned in my original ticket, I think deterministic optimizers can help expand the pytorch horizon by offering high-performance implementations of widely-used tools from libraries like SciPy and MATLAB. It would seem to fit well with other ongoing pytorch projects, such as the `linalg` module.

**Comment by mruberry** [comment]
Makes sense, @rfeinman. @vincentqb, what are your thoughts?

**Comment by rfeinman** [comment]
I put together a write-up of the proposed API [here](https://pytorch-minimize.readthedocs.io/en/latest/api/index.html). It includes a summary of both the functional and object-oriented APIs that I've mentioned. Would be curious to hear your thoughts.

**Comment by tvercaut** [comment]
It would be great to support such a functionality in pytorch as this is a core functionality available in scipy and there is no actively maintained third-party pytorch library for it.

This also echos related feature requests for Levenberg-Marquardt, Non-linear Conjugate Gradient and the like:
#83529 #51407 #80553 

Other related libraries:
https://github.com/hahnec/torchimize

To some extent, one can already rely on [LBFGS](https://pytorch.org/docs/stable/generated/torch.optim.LBFGS.html) but this may not always be the best option and bound constraints have not been implemented yet. That is, there is no L-BFGS-B in pytorch:
#22281 #6564

**Comment by tvercaut** [comment]
I have just stumbled onto [Theseus](https://github.com/facebookresearch/theseus) which addresses some of the needs discussed here: namely unconstrained non-linear-least squares.
